src/cache.c

```diff
@@ -214,66 +264,77 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     }
 
     /* get the entry to be replaced if cache is full */
-    if (!replaced_entry && cache->list_size == cache->capacity) {
-        for (int i = 0; i < THRESHOLD; i++) {
-            if (list_empty(cache->lists[i]))
-                continue;
-            replaced_entry =
-                list_last_entry(cache->lists[i], lfu_entry_t, list);
-            break;
-        }
+    if (!replaced_entry && cache->size == cache->capacity) {
+        replaced_entry = list_last_entry(&cache->list, cache_entry_t, list);
         assert(replaced_entry);
     }
 
     void *replaced_value = NULL;
     if (replaced_entry) {
+        assert(replaced_entry->alive);
+
         replaced_value = replaced_entry->value;
+        replaced_entry->alive = false;
         list_del_init(&replaced_entry->list);
-        hlist_del_init(&replaced_entry->ht_list);
-        mpool_free(cache_mp, replaced_entry);
-        cache->list_size--;
+        cache->size--;
+        list_add(&replaced_entry->list, &cache->ghost_list);
+        cache->ghost_list_size++;
     }
 
-    lfu_entry_t *new_entry = mpool_alloc(cache_mp);
+    cache_entry_t *new_entry = calloc(1, sizeof(cache_entry_t));
     INIT_LIST_HEAD(&new_entry->list);
     INIT_HLIST_NODE(&new_entry->ht_list);
     new_entry->key = key;
     new_entry->value = value;
-    new_entry->frequency = 0;
-    list_add(&new_entry->list, cache->lists[new_entry->frequency++]);
+    new_entry->alive = true;
+
+    if (!revived_entry) {
+        new_entry->frequency = 1;
+    } else {
+        new_entry->frequency = revived_entry->frequency + 1;
+        hlist_del_init(&revived_entry->ht_list);
+        list_del_init(&revived_entry->list);
+        cache->ghost_list_size--;
+        free(revived_entry);
+    }
+
+    list_add(&new_entry->list, &cache->list);
     hlist_add_head(&new_entry->ht_list,
-                   &cache->map->ht_list_head[cache_hash(key)]);
-    cache->list_size++;
+                   &cache->map.ht_list_head[cache_hash(key)]);
 
-    assert(cache->list_size <= cache->capacity);
+    cache->size++;
+
+    cache_ghost_list_update(cache);
+
+    assert(cache->size <= cache->capacity);
+    assert(cache->ghost_list_size <= cache->capacity);
     return replaced_value;
 }
 
 void cache_free(cache_t *cache)
 {
-    for (int i = 0; i < THRESHOLD; i++)
-        free(cache->lists[i]);
-    mpool_destroy(cache_mp);
-    free(cache->map->ht_list_head);
-    free(cache->map);
+    free(cache->map.ht_list_head);
     free(cache);
 }
 
 uint32_t cache_freq(const struct cache *cache, uint32_t key)
 {
-    if (!cache->capacity ||
-        hlist_empty(&cache->map->ht_list_head[cache_hash(key)]))
+    if (unlikely(!cache->capacity))
```

Comment:
Shouldn't it be checking the `size` since the `capacity` is fixed after initialization but `size` changes during runtime?

Reply:
If the cache capacirt is set to zero, we could and should alway return 0/NULL for all cases.

Reply:
> If the cache capacity is set to zero, we could and should alway early return with 0 for all cases.

Got it. The `size` can be checked by `hlist_empty`.